internal/ls/testsetup.go

```diff
@@ -0,0 +1,179 @@
+package ls
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/microsoft/typescript-go/internal/core"
+)
+
+type markerRange struct {
+	core.TextRange
+	filename string
+	position int
+	data     string
+}
+
+type Marker struct {
+	Filename string
+	Position int
+	Name     string
+}
+
+type TestData struct {
+	Files           []TestFileInfo
+	MarkerPositions map[string]Marker
+	//markers         []marker
+	/**
+	 * Inserted in source files by surrounding desired text
+	 * in a range with `[|` and `|]`. For example,
+	 *
+	 * [|text in range|]
+	 *
+	 * is a range with `text in range` "selected".
+	 */
+	Ranges markerRange
+}
+
+func ParseTestData(basePath string, contents string, fileName string) TestData {
+	// List of all the subfiles we've parsed out
+	files := []TestFileInfo{}
+
+	// Split up the input file by line
+	lines := strings.Split(contents, "\n")
+	currentFileContent := ""
+
+	for _, line := range lines {
+		if len(line) > 0 && line[len(line)-1] == '\r' {
+			line = line[:len(line)-1]
+		}
+		if currentFileContent == "" {
+			currentFileContent = line
+		} else {
+			currentFileContent += "\n" + line
+		}
+	}
+
+	if currentFileContent == "" {
+		return TestData{}
+	}
+	markerPositions := make(map[string]Marker)
+	markers := []Marker{}
+
+	// If we have multiple files, then parseFileContent needs to be called for each file. This will be achieved by creating a `nextFile()`` func that will call `parseFileContent()`` for each file.
+	testFileInfo := parseFileContent(currentFileContent, fileName, &markerPositions, &markers)
+	files = append(files, testFileInfo)
+
+	return TestData{
+		Files:           files,
+		MarkerPositions: markerPositions,
+		// markers:         markers,
+		Ranges: markerRange{},
+	}
+}
+
+type locationInformation struct {
+	position       int
+	sourcePosition int
+	sourceLine     int
+	sourceColumn   int
+}
+
+type TestFileInfo struct { // for FourSlashFile
+	Filename string
+	// The contents of the file (with markers, etc stripped out)
+	Content string
+}
+
+func parseFileContent(content string, filename string, markerMap *map[string]Marker, markers *[]Marker) TestFileInfo {
```

Comment:
Pointers to maps and slices look fishy

Reply:
Fixed the pointer to map, but the pointer to slice is needed because we append to the slice.